tests/stats_test.cpp

```diff
@@ -0,0 +1,390 @@
+}
+
+static void test_event_stats_with_time_dependency()
+{
+	const size_t default_size = 100;
+	const time_t default_time = time(nullptr);
+	stats_t stats(EVENTS_SIZE, PERIOD_IN_SECONDS);
+	stats_t stats2(EVENTS_SIZE, PERIOD_IN_SECONDS);
+	stats_t stats3(EVENTS_SIZE, PERIOD_IN_SECONDS);
+	std::vector<test_event> result, top_events;
+
+	// old events (outside observable period of time) are omitted
+	const size_t long_period = 10 * PERIOD_IN_SECONDS;
+	for(size_t i = 0; i < long_period; ++i) {
+		test_event e{"same", default_size, 1., default_time + static_cast<time_t>(i)};
+		stats.add_event(e, e.get_time());
+	}
+
+	stats.get_top(TOP_LENGTH, default_time + long_period, result);
+	BOOST_REQUIRE_EQUAL(result.size(), 1);
+	BOOST_CHECK(result.back().get_weight() <= PERIOD_IN_SECONDS * default_size);
+
+	// weight of event must be smaller after significant period of silence
+	test_event e{"same", default_size, 1., default_time + static_cast<time_t>(long_period + PERIOD_IN_SECONDS / 2)};
+	stats.add_event(e, e.get_time());
+	result.clear();	
+	stats.get_top(TOP_LENGTH, e.get_time(), result);
+	BOOST_REQUIRE_EQUAL(result.size(), 1);
+	BOOST_CHECK(result.back().get_weight() <= PERIOD_IN_SECONDS * default_size / 2);
+
+	// small key with regular frequent access must popup in top stats among heavier keys with single access
+	for(int i = 0; i < 3 * TOP_LENGTH; ++i) {
+		size_t size = i % 2 ? default_size : default_size * 10;
+		std::string key = i % 2 ? std::string("sum") : std::to_string(static_cast<long long>(i));
+
+		test_event e{key, size, 1., default_time + i};
+		stats2.add_event(e, e.get_time());
+	}
+
+	result.clear();	
+	stats2.get_top(TOP_LENGTH, default_time + 3 * TOP_LENGTH, result);
+	BOOST_REQUIRE_EQUAL(result.size(), TOP_LENGTH);
+	BOOST_REQUIRE_EQUAL(*result.front().get_key(), "sum");
+	BOOST_CHECK(result.front().get_weight() > result.back().get_weight());
+	BOOST_CHECK(result.front().get_frequency() > result.back().get_frequency());
+
+	// if keys have same size and time access pattern, then keys with more frequent access must be in top
+	for(int i = 0; i < 5 * TOP_LENGTH; ++i) {
+		std::string key = std::to_string(static_cast<long long>(i));
+		for(int j = 0; j < i+1; ++j) {
+			test_event e{key, default_size, 1., default_time + j};
+			stats3.add_event(e, e.get_time());
+		}
+		if (i >= 4 * TOP_LENGTH) {
+			test_event e{key, default_size, 1., default_time};
+			top_events.push_back(e);
+		}
+	}
+
+	result.clear();
+	stats3.get_top(TOP_LENGTH, default_time + 5 * TOP_LENGTH, result);
+	BOOST_REQUIRE_EQUAL(result.size(), TOP_LENGTH);
+	BOOST_REQUIRE_EQUAL(events_symmetric_diff(top_events, result), 0);
+}
+
+bool register_tests(test_suite *suite, node n)
+{
+	(void) n;
+	ELLIPTICS_TEST_CASE_NOARGS(test_top_provider_existance);
+	ELLIPTICS_TEST_CASE_NOARGS(test_event_stats_boundary_conditions);
+	ELLIPTICS_TEST_CASE_NOARGS(test_event_stats_no_time_dependency);
+	ELLIPTICS_TEST_CASE_NOARGS(test_event_stats_with_time_dependency);
```

Comment:
>describe it with more information about design of the tests

event_stats - is a core abstract data type used in top keys statistics logic. If event_stats logic is correct, then top keys statistics will be correct too.
event_stats ADT provides statistics logic for ANY type, that supports methods {set/get}_weight, {set/get}_time, {set/get}_frequency.

test_top_provider_existence - if "top" section in "monitor" section in config exists and well-defined, then top_provider object must exists.

test_event_stats_boundary_conditions - various boundary conditions test, e.g. inserting more events than event_stats::m_max_events (== EVENTS_SIZE) will fit into event_stats max capacity (get_top() should never return more than m_max_events events).

test_event_stats_no_time_dependency - test that checks (binary) heap properties of event_stats ADT, e.g. monotonically increment event's weight within events_limit and then see if heaviest event occurs in get_top() result. All inserted events have same time of occurance.

test_event_stats_with_time_dependency - checks that event's weight and event's existence in event_stats container depends on time as it was designed (satisfy heap property and "old" events are not returned by get_top()).

I will add more comments in test_stats.cpp in next (fixing) PR.